internal/testutils/broker.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	uiLayoutInfo, err := b.broker.SelectAuthenticationMode(context.Background(), sessionID, authenticationModeName)
+	if err != nil {
+		return nil, dbus.MakeFailedError(err)
+	}
+	return uiLayoutInfo, nil
+}
+
+// IsAuthorized calls the broker mock corresponding method and returns the result or an error if requested.
+func (b *BrokerBusMock) IsAuthorized(sessionID, authenticationData string) (access, infoUser string, dbusErr *dbus.Error) {
+	if sessionID == "IA_bus_error" {
+		return "", "", dbus.MakeFailedError(errors.New("erroring out as requested"))
+	}
+
+	access, infoUser, err := b.broker.IsAuthorized(context.Background(), sessionID, authenticationData)
+	if err != nil {
+		return "", "", dbus.MakeFailedError(err)
+	}
+	return access, infoUser, nil
+}
+
+// EndSession calls the broker mock corresponding method and returns the result or an error if requested.
+func (b *BrokerBusMock) EndSession(sessionID string) (dbusErr *dbus.Error) {
+	if sessionID == "ES_bus_error" {
+		return dbus.MakeFailedError(errors.New("erroring out as requested"))
+	}
+
+	err := b.broker.EndSession(context.Background(), sessionID)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	return nil
+}
+
+// CancelIsAuthorized calls the broker mock corresponding method.
+func (b *BrokerBusMock) CancelIsAuthorized(sessionID string) (dbusErr *dbus.Error) {
+	b.broker.CancelIsAuthorized(context.Background(), sessionID)
+	return nil
+}
+
+/* BrokerMock implementation */
+
+// NewSession returns default values to be used in tests or an error if requested.
+func (b *BrokerMock) NewSession(ctx context.Context, username, lang string) (sessionID, encryptionKey string, err error) {
+	if username == "NS_error" {
+		return "", "", errors.New("erroring out as requested")
+	}
+	return "id", "encryption", nil
+}
+
+// GetAuthenticationModes returns default values to be used in tests or an error if requested.
+func (b *BrokerMock) GetAuthenticationModes(ctx context.Context, sessionID string, supportedUILayouts []map[string]string) (authenticationModes []map[string]string, err error) {
+	switch sessionID {
+	case "GAM_invalid":
+		return []map[string]string{
+			{"invalid": "invalid"},
+		}, nil
+
+	case "GAM_empty":
+		return nil, nil
+
+	case "GAM_error":
+		return nil, errors.New("erroring out as requested")
+	}
+
+	return []map[string]string{
+		{"name": "mode1", "label": "Mode 1"},
+	}, nil
+}
+
+// SelectAuthenticationMode returns default values to be used in tests or an error if requested.
+func (b *BrokerMock) SelectAuthenticationMode(ctx context.Context, sessionID, authenticationModeName string) (uiLayoutInfo map[string]string, err error) {
+	switch sessionID {
+	case "SAM_invalid_layout":
+		return map[string]string{
+			"invalid": "invalid",
+		}, nil
+
+	case "SAM_no_layout":
+		return nil, nil
+
+	case "SAM_error":
+		return nil, errors.New("erroring out as requested")
+	}
+
+	return map[string]string{
```

Comment:
same, I was forseeing more use case, like returning invalid type, invalid fields for a given type, to ensure that all the support cases are covered.

Reply:
Also, we can imagine other cases with fields that are ignored, and so on.

Reply:
To be honest, I didn't go too deep on this because I wasn't sure we were 100% certain of how the end result would be, so I treated it more generally and would go back on it once it was defined.